server/src/main/java/io/crate/expression/scalar/geo/DistanceFunction.java

```diff
@@ -120,4 +132,107 @@ public Symbol normalizeSymbol(Function symbol, TransactionContext txnCtx, NodeCo
         }
         return symbol;
     }
+
+    @Override
+    public Query toQuery(Function parent, Function inner, Context context) {
+        assert inner.name().equals(DistanceFunction.NAME) :
+            "function must be " + DistanceFunction.NAME;
+
+        //                 ┌─► pointLiteral
+        //                 │
+        // distance(p, 'POINT (1 2)') = 10
+        //          │                   │
+        //          └► pointRef         │
+        // ^^^^^^^^^^^^^^^^^^^^^^^^^^   └► parentRhs
+        //          inner
+        //
+        // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+        //              parent
+        //
+        List<Symbol> innerArgs = inner.arguments();
+        if (!(innerArgs.get(0) instanceof Reference pointRef && innerArgs.get(1) instanceof Literal<?> pointLiteral)) {
+            // can't use distance filter without literal, fallback to genericFunction
+            return null;
+        }
+        List<Symbol> parentArgs = parent.arguments();
+        if (!(parentArgs.get(1) instanceof Literal<?> parentRhs)) {
+            return null;
```

Comment:
maybe we keep the (a similar) comment here as well?
`// must be something like cmp(distance(..), non-literal) - fallback to genericFunction`

Reply:
Done :+1: